### PR TITLE
More dependency verification usability improvements

### PR DESCRIPTION
**Pull request by melix**
Dogfooding FTW!

See commits for details

### CODE REVIEW

**Comment by melix** [comment]
@ljacomet this is RFR

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java [hunk 1]
```diff
@@ -79,12 +81,19 @@ public DependencyVerificationOverride dependencyVerificationOverride(BuildOperat
...
             if (verificationsFile.exists()) {
+                if (startParameter.getDependencyVerificationMode() == DependencyVerificationMode.OFF) {
+                    return DependencyVerificationOverride.NO_VERIFICATION;
+                }
                 SingleMessageLogger.incubatingFeatureUsed("Dependency verification");
-                return new ChecksumVerificationOverride(buildOperationExecutor, verificationsFile, checksumService);
+                try {
+                    return new ChecksumVerificationOverride(buildOperationExecutor, verificationsFile, checksumService, startParameter.getDependencyVerificationMode());
+                } catch (Exception e) {
+                    return new FailureVerificationOverride(e, verificationsFile);
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java]: I am not sure I understand the real benefit of this change.
It effectively enables to run `help` or `tasks` but not much more.
Is there something else I am missing for this feature?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java [hunk 2]
```diff
@@ -145,6 +152,7 @@ private void maybeReadExistingFile() throws FileNotFoundException {
     private void writeEntriesSerially() {
         entriesToBeWritten.stream()
             .sorted()
+            .filter(entry -> !isTrustedArtifact(entry.id))
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java]: I wonder if that should not be a fully separate option, to allow controlling trusted artifacts and their presence in the file independently.

Also with these options, the content of a verification file can vary.
I think we should document the minimal file you could create by hand, with your configuration, before populating the file with entries the first time.
Makes sense? (and of course for a later PR)

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java (hunk 1 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java]: It's more than this. Without this change, Gradle fails to initialize: the `DependencyManagementBuildScopeServices` fail to create their dependent services, so the error that is thrown to the user is not nice. With this change, the error is captured and thrown _after_ initialization.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java (hunk 2 above)
Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java]: We could split the file in 2, yes. But at this stage I'm not convinced it has a major advantage.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/StartParameterResolutionOverride.java]: Ah ok, makes sense then! 👍

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java (hunk 2 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java]: Not saying we should split the file in 2, just making sure we document that you can have something like:
```
<?xml version="1.0" encoding="UTF-8"?>
<verification-metadata>
   <configuration>
      <verify-metadata>false</verify-metadata>
      <trusted-artifacts/>
   </configuration>
</verification-metadata>
```
And then  metadata and trusted artifacts would be ignored from the first generation of the file.

Comment (melix) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/verification/WriteDependencyVerificationFile.java]: Right, I don't plan to start documentation until the dogfooding branch passes.

**Pull request merged**